### ISSUE TITLE
Downloads either fail or the app just outright freezes and needs to be killed in Task Manager

### ISSUE THREAD

**Issue by michael-zock** [rmcrackan/Libation#375]
I just installed the program without problems and was able tos et up my account, but once I try to actually get my audiobooks, the problem always eventually freezes and the task needs to be killed (just killing it via the task bar still leaves the process behind).

In the example of the attached log, the "included in my subscription" stuff outright fails and the first chapter of a purchased book at least gets the progress bar about 10-15 percent of the way before once again the inevitable happens.
If I hadn't activated the option to just retry a download later it probably would be an even worse experience.

System is Windows 11 version 21H2 with current updates.

Hope someone can make sense of this since I'm at my wits' end, This tool freezes, OpenAudible is no longer open but instead a cash grab and Book Lib Connect also fails to download anything. I'm half a step away from giving up on Audible and raise the black flag like in the old days.

**Thanks in advance!**

[Log202209.log](https://github.com/rmcrackan/Libation/files/9638585/Log202209.log)

**Comment by rmcrackan** [comment]
Eek. So sorry this happened to you. Much thanks for the log file and for giving us a chance to make this right.

I feel like we had this problem in the past, possibly even for this title (uk: Aliens꞉ Bug Hunt [B07D2HJNS3]) and I thought we fixed it. Does it help to delete all of your files in these folders and try again?

* `C:\Users\michael\AppData\Local\Temp\Libation\DownloadsInProgress`
* `C:\Users\michael\AppData\Local\Temp\Libation\DecryptInProgress`

@Mbucari Do you recall what we did?

**Comment by michael-zock** [comment]
Thanks!
I emptied  both folders and unfortunately it once again happens straight away once the library is loaded and it tries to download something for later decryption (progress bar appears below the item and that's the last thing that happens).
[Log202209.log](https://github.com/rmcrackan/Libation/files/9638654/Log202209.log)

**Comment by rmcrackan** [comment]
Can you turn off settings for auto scan and auto download? I think these are happening at the same time which is muddling the logs. Please turn those off, restart libation, manually try again, and upload the new log.

**Comment by michael-zock** [comment]
Thanks again for looking into it. Here's the steps since then:
1. Changed the two settings.
2. Closed the program.
3. Deleted the old log file.
4. Opened the program again.
5. Scrolled down to the first regular book (added without a credit).
6. Clicked on the item on the left.
As usual, as soon as the ETA time gets its first value, the program stops reacting to mouse inputs and the progress bar doesn't move either.

[Log202209.log](https://github.com/rmcrackan/Libation/files/9638850/Log202209.log)

**Comment by Mbucari** [comment]
@rmcrackan The content license denials are strange. Usually it's just 1 or 2 odd titles, and the problem resolves itself in short order. But here licenses are being denied for every single book. Is this maybe an API change?

See my email for the app hang.

**Comment by michael-zock** [comment]
Weird thing is, I tried it again today and a few more of them worked, We're only talking about a success rate of about 5% of the titles, but still...

**Comment by rmcrackan** [comment]
@michael-zock Have you been downloading a lot of free titles lately? That could explain this latest development. Audible is starting to limit how many free titles a user can download in a day.

**Comment by rmcrackan** [comment]
@Mbucari Every book? Or every free/"Plus" book? If the latter, I suspect we're just hitting a daily limit.

**Comment by Mbucari** [comment]
@rmcrackan 

> Every book?

All 53 books [Log202209.log](https://github.com/rmcrackan/Libation/files/9638585/Log202209.log) (except for the one that had an SSL error) were denied a content license. None succeeded. IDK if they were all Plus books.

**Comment by michael-zock** [comment]
A handful of free books now work, the paid ones sail through. The "free" ones initially didn't work at all.

**Comment by rmcrackan** [comment]
Content license denied is consistent with other people who've experienced audible's plus-download daily limit. If progress can be made every 1-2 days on those before they lock-up again, I'm fine with calling this case closed.

The parts I find weird are [1] why would this cause the UI to freeze? [2] Why would it cause the error "The specified data could not be decrypted." ?

**Comment by michael-zock** [comment]
Yes, as long as some progress can be made by waiting and not giving up it's the freezing, to the point where not even killing the process via the task bar (only via the task manager's "Details" tab) is able to fully get it out of the way for a new try that poses the deeper problem.

**Comment by Mbucari** [comment]
> [1] why would this cause the UI to freeze?

There's a bug in aax clean I will cause the collar to hang if the input stream is terminated before it has finished reading the file.

> [2] Why would it cause the error "The specified data could not be decrypted." ?

It's the other way around. Some SSL error is causing the HTTP client Network stream to close. The network stream closing early is what causes the app to hang because of the bug in [1].

**Comment by rmcrackan** [comment]
Sounds like the aax clean bug fix will solve the freezing, the plus catalog daily limit is causing that issue and incremental progress can be made over time, and the network stream closing early is something that just calls for a retry. Am I missing anything?

**Comment by Mbucari** [comment]
> Am I missing anything?

I think that's it. Maybe just leave this issue open until we managed to fix that bug.

**Comment by rmcrackan** [comment]
@Mbucari Thanks for the fix!

@michael-zock Please try [v8.4.6](https://github.com/rmcrackan/Libation/releases/tag/v8.4.6) and let me know if the problem persists.

**Comment by michael-zock** [comment]
Thanks!
It's still running into problems, but at least it now fails gracefully so I can retry later.
[Log202209.log](https://github.com/rmcrackan/Libation/files/9667989/Log202209.log)

**Comment by rmcrackan** [comment]
Are you still making some progress on plus titles and then it breaks again on a plus title?

**Comment by michael-zock** [comment]
Some, it feels like a few of the "free" ones have been unlocked since this problem originally popped up.

**Quick question:** When items fail, the program says it will retry later. Is there some timer involved after which Libation will automatically do another attempt or do I need to clear the queue and trigger another run manually?

**Comment by Mbucari** [comment]
Using the url from your log file, I was able to download _Alien꞉ Out of the Shadows꞉ An Audible Original Drama_. 
That means that you were given a valid link, but your computer is unable to make a stable connection to cloudfront.

I'm afraid this is a problem with your network or local ISP.

**Comment by michael-zock** [comment]
Thanks, yes that title now completed on it's third try. I'll have to look into connection problems

**Comment by michael-zock** [comment]
Still leaves the question, is there an automated way to retry later or do I have to clear the log and launch another run of the list manually later? The "Error, will retry later" message is kinda ambiguous about that.
Something timer-based would mean that someone in my situation could just let the app running through the night instead of having to recheck repeatedly.

**Comment by rmcrackan** [comment]
Sorry for the ambiguity. It means that the book will be retried on the next attempt to download all books. As opposed to not trying again ever -- which is preferable when there's something genuinely wrong with the title and you don't want it throwing errors again for a known bad book.

I'm not sure how this will act with the errors you're getting: there are the options (Settings, Import library tab) to automatically run period scans and to auto download new books. (It says Auto download episodes but I think I must have goofed when I wrote that -- I'll check soon.) In theory this will keep trying. I'm not sure how these frequent errors will affect it.

**Comment by rmcrackan** [comment]
Confirmed: just poor wording. It will be fixed in the next release

**Comment by michael-zock** [comment]
Thanks again for your help and for looking into the auto retry at a later date. I've enabled it so I won't forget.

**Comment by rmcrackan** [comment]
It seems like the Libation issues are fixed and what remains is network and/or audible being a jerk so I'm going to close this issue. If you reply here we'll still get notices though.

**Comment by michael-zock** [comment]
@rmcrackan Little update in case anyone else ever runs into this: 

- Libation is now able to download all titles (free and paid) after the free titles have been unlocked by Audible in batches as suspected, except for one that's understandable since it's just pre-booked for its release next month.
- Turns out, the problem was related to the no-name USB WIFI module on my desktop. I borrowed a different one to test and although it was slower (didn't have 5 GHz capability even though it claims on the label), downloads now worked uninterrupted. As a result, I bought a decent one that has an antenna for better reception and can work on both frequency bands and ever since then everything works as intended after you two fixed that "freezing" problem.

Thanks again for everything and I hope this will be able to help others in the future.

**Comment by rmcrackan** [comment]
It's a cliche among software engineers to blame things on the hardware. Holy crap -- a bug that's legit a hardware problem. That just made my whole week!
